File: src/main/java/net/Chidoziealways/everythingjapanese/chakra/ChakraSyncPacket.java
package net.Chidoziealways.everythingjapanese.chakra;

import net.Chidoziealways.everythingjapanese.capabilities.ModCapabilities;
import net.minecraft.client.Minecraft;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraftforge.event.network.CustomPayloadEvent;

public record ChakraSyncPacket(int chakra) {

    public static void encode(ChakraSyncPacket msg, FriendlyByteBuf buffer) {
        buffer.writeInt(msg.chakra);
    }

    public static ChakraSyncPacket decode(FriendlyByteBuf buffer) {
        return new ChakraSyncPacket(buffer.readInt());
    }

    public static void handle(ChakraSyncPacket packet, CustomPayloadEvent.Context context) {
        context.enqueueWork(() -> {
            assert Minecraft.getInstance().player != null;
            Minecraft.getInstance().player.getCapability(ModCapabilities.CHAKRA_CAPABILITY)
                    .ifPresent(iChakra -> iChakra.setChakra(packet.chakra));
        });
        context.setPacketHandled(true);
    }
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/chakra/ChakraSyncPacket.kt
package net.Chidoziealways.everythingjapanese.chakra

import net.Chidoziealways.everythingjapanese.JAPANESE_MOD_ID
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier

data class ChakraSyncPacket(@JvmField val chakra: Float, val maxChakra: Int): CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> {
        return TYPE
    }

    companion object {
        val ID = Identifier.fromNamespaceAndPath(JAPANESE_MOD_ID, "chakra_sync")
        val TYPE = CustomPacketPayload.Type<ChakraSyncPacket>(ID)

        val CODEC: StreamCodec<FriendlyByteBuf, ChakraSyncPacket> =
            StreamCodec.of(
                { buf, value ->
                    buf.writeFloat(value.chakra)
                    buf.writeInt(value.maxChakra)
                },
                { buf -> ChakraSyncPacket(buf.readFloat(), buf.readInt()) }
            )
    }
}
